### ISSUE TITLE
TriangleCollector::visitTrianglesConstrained - visits null vertex

### ISSUE THREAD

**Issue by jandam** [gwlucastrig/Tinfour#33]
TriangleCollector::visitTrianglesConstrained in some cases visits triangle with null vertex. Integrity check is valid.
  For now no action is required. First of all I have to check validity of input data.

**Comment by gwlucastrig** [comment]
I looked through the code.  I did not find anything that was unambiguously the cause of your problem.

One case I can imagine where this might happen is if you defined a constrained region with a single clockwise polygon...  Thus the polygon would define a set of data with a hole in the middle.  All triangles exterior to the polygon would be valid members of the constrained region.  Unfortunately, that might include ghost triangles (triangles with null vertices).   

In the current implementation, TriangleCollector does not test for null vertices before calling accept().  Perhaps it should.

I will think about this problem a bit more.

**Comment by jandam** [comment]
I added test for null vertices. It works and I didn't see any issues. But still I have to make some investigation about my data.

**Comment by gwlucastrig** [comment]
I've looked at this problem some more.   I will be adding a test to the TriangleCollector to screen out cases where the mesh traversal logic encounters triangles that contain null vertices. 

Recall that Tinfour takes the well-known approach of defining a "ghost vertex" and a set of "ghost triangles" that are exterior to the convex hull which encloses the set of vertices from which the triangulated mesh is built.  While other Delaunay Triangulation implementations define the ghost vertex using alternate conventions, Tinfour treats it as a null reference.

In the scenario I described in a post above above, a clockwise polygon defines a region of all triangles exterior to the polygon.  In such a case, the TriangleCollector's mesh-traversal logic does, in fact, visit ghost triangles (those containing the ghost vertex, which Tinfour represents using a null).  There are cases where the traversal logic must visit the ghost triangles in order to ensure that it collects all valid triangles (the case where some, but not all, of the edges of the clockwise polygon lie on the convex hull of the set of vertices).   Thus there is no way to pre-screen the  triangles containing a null vertex prior to the traveral.  The only way to prevent them from being passed to the TriangleCollector's accept() method is by testing them right before accept() is called.  This approach adds undesirable overhead to the processing, but is unavoidable.  

A code update will be posted some time this weekend.
